hoomd/Autotuner.h

```diff
@@ -24,227 +34,620 @@
+            std::ostringstream s;
+            s << "Error setting autotuner parameter " << m_name << ". Got " << n_params
+              << " parameters, "
+              << "expected " << n_dimensions << ".";
+            throw std::runtime_error(s.str());
+            }
 
-            // if not complete, issue a warning
-            if (!isComplete())
-                {
-                m_exec_conf->msg->notice(2) << "Disabling Autotuner " << m_name
-                                            << " before initial scan completed!" << std::endl;
-                }
-            else
-                {
-                // ensure that we are in the idle state and have an up to date optimal parameter
-                m_current_element = 0;
-                m_state = IDLE;
-                m_current_param = computeOptimalParameter();
-                }
+        std::array<unsigned int, n_dimensions> cpp_param;
+        for (size_t i = 0; i < n_dimensions; i++)
+            {
+            cpp_param[i] = pybind11::cast<unsigned int>(parameter[i]);
             }
-        else
+
+        /// Check that the parameter is valid.
+        auto found = std::find(m_parameters.begin(), m_parameters.end(), cpp_param);
+        if (found == m_parameters.end())
             {
-            m_exec_conf->msg->notice(6) << "Enable Autotuner " << m_name << std::endl;
+            std::ostringstream s;
+            s << "Error setting autotuner parameter " << m_name << ". " << formatParam(cpp_param)
+              << " is not a valid parameter.";
+            throw std::runtime_error(s.str());
             }
-        }
 
-    //! Test if initial sampling is complete
-    /*! \returns true if the initial sampling run is complete
-     */
-    bool isComplete()
-        {
-        if (m_state != STARTUP)
-            return true;
-        else
-            return false;
+        m_current_param = cpp_param;
+        m_state = IDLE;
+        m_current_sample = 0;
+
+        m_exec_conf->msg->notice(4) << "Autotuner " << m_name << " setting user-defined parameter "
+                                    << formatParam(cpp_param) << std::endl;
         }
 
-    //! Get the sampling period
-    unsigned int getPeriod() const
+    static std::string formatParam(const std::array<unsigned int, n_dimensions>& p)
         {
-        return m_period;
+        std::ostringstream s;
+        s << "(";
+        for (auto v : p)
+            {
+            s << v << ",";
+            }
+        s << ")";
+        return s.str();
         }
 
-    //! Change the sampling period
-    /*! \param period New period to set
+    /// Test if the scan is complete.
+    /*! \returns true when autotuning is complete.
      */
-    void setPeriod(unsigned int period)
+    virtual bool isComplete()
         {
-        m_exec_conf->msg->notice(6)
-            << "Set Autotuner " << m_name << " period = " << period << std::endl;
-        m_period = period;
+        if (m_state != SCANNING)
```

Comment:
This is intentional. I want tuners that are created, but`begin` is never called to be "complete". Otherwise, they would prevent the `Autotuned` object as a whole from ever registering as complete.